### DISCUSSION TITLE
Logic/Backend and UI bindings - missing notifications and read only problems

### DISCUSSION THREAD

**Discussion by jahnf** [slint-ui/slint#4199] [Q&A] [answered]
Hello, I am experimenting with slint and came across a generic issue I have when separating Application/Backend logic from the UI.

The way to go seems as in the example a global component (which sadly cannot contain sub components).

I have multiple issues, so I have to resort to hacks like my simple example below.
1.  If I have an out only property that is meant to be only readable by the UI, this property cannot be written by the backend, I have to resort to the hack below.
2.  If I want a property that can be read and written by the UI, there is no way (at least that I know of) how the backend can be notified of a change, so I have to resort to the hack below.

```js
export global BackendImpl {
  // --- actual property values - ohly to utilized within glue code.
  in-out property <int> my-value;
}

export global Backend  {
  // --- properties for bindings within the UI
  out property <int> my-value: BackendImpl.my-value;
  // --- functions (if applicable) to  set properties from within the UI
  // --- this guarantees that the glue code UI <> Backend is (can be) notified
  pure callback write-my-value(int);
}
```

**Is there a better solution?** this seems so 'hacky'.

**Comment by ogoffart** [comment, marked as answer]
This is the way I recommend to do it, yes.

I'd even put a default implementation so it works in the preview without the backend:

```js 
export global Backend  {
  in-out property <int> my-value;
  pure callback write-my-value(int);
  // default implementation that can be overridden by the backend
  write-my-value(x) => { my-value = x; }
}
```

**Reply by jahnf** [reply]
By the way the default implementation does not work for me 
I get an error  `Assignment in a pure context`

**Reply by ogoffart** [reply]
ah sorry.  The callback don't need to be pure:

```
export global Backend  {
  in-out property <int> my-value;
  callback write-my-value(int); // removed "pure"
  write-my-value(x) => { my-value = x; }
}
```
